Firestore/src/FirestoreSessionHandler.php

```diff
@@ -0,0 +1,265 @@
+ * ```
+ * @see http://php.net/manual/en/class.sessionhandlerinterface.php SessionHandlerInterface
+ */
+class FirestoreSessionHandler implements SessionHandlerInterface
+{
+    /* @var int */
+    private $gcLimit;
+    /* @var FirestoreClient */
+    private $firestore;
+    /* @var CollectionReference */
+    private $collection;
+
+    /**
+     * Create a custom session handler backed by Cloud Firestore.
+     *
+     * @param FirestoreClient $firestore Firestore client.
+     * @param int $gcLimit [optional] A number of entities to delete in the
+     *        garbage collection.  Defaults to 0 which means it does nothing.
+     *        The value larger than 1000 will be cut down to 1000.
+     */
+    public function __construct(
+        FirestoreClient $firestore,
+        $gcLimit = 0
+    ) {
+        $this->firestore = $firestore;
+        // Cut down to 1000
+        $this->gcLimit = min($gcLimit, 1000);
+    }
+
+    /**
+     * Start a session, by getting the Firebase collection from $savePath.
+     *
+     * @param string $savePath The value of `session.save_path` setting will be
+     *        used here as the Firestore collection ID.
+     * @param string $sessionName The value of `session.name` setting will be
+     *        used here as the Firestore document ID prefix (ex: "PHPSESSID:").
+     * @return bool
+     */
+    public function open($savePath, $sessionName)
+    {
+        if (false !== strpos($savePath, '/')) {
+            throw new InvalidArgumentException(
+                sprintf('The given save_path "%s" not allowed', $savePath)
+            );
+        }
+        $this->sessionName = $sessionName;
+        $this->collection = $this->firestore->collection($savePath);
+        return true;
+    }
+
+    /**
+     * Just return true for this implementation.
+     */
+    public function close()
+    {
+        return true;
+    }
+
+    /**
+     * Read the session data from Cloud Firestore.
+     *
+     * @param string $id Identifier used for the session.
+     * @return string
+     */
+    public function read($id)
+    {
+        try {
+            $docRef = $this->collection->document($this->formatId($id));
+            $snapshot = $docRef->snapshot();
+            if ($snapshot->exists() && isset($snapshot['data'])) {
+                return $snapshot->get('data');
+            }
+        } catch (Exception $e) {
+            trigger_error(
+                sprintf('Firestore lookup failed: %s', $e->getMessage()),
+                E_USER_WARNING
+            );
+        }
+        return '';
+    }
+
+    /**
+     * Write the session data to Cloud Firestore.
+     *
+     * @param string $id Identifier used for the session.
+     * @param string $data The session data to write to the Firestore document.
+     * @return bool
+     */
+    public function write($id, $data)
+    {
+        try {
+            $docRef = $this->collection->document($this->formatId($id));
```

Comment:
Should this and other service calls (especially writes) be wrapped in `FirestoreClient::runTransaction()` to handle retry of race conditions?